File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples/VB/WPFTemplateApp/WPFTemplateApp/Common/ViewModelBase.vb
Public Enum UIStateMode
  Normal
  Edit
  Add
  Cancel
End Enum

Public Class ViewModelBase
  Inherits ObjectBase
#Region "Private Variables"
  Private _IsSaveEnabled As Boolean = False
  Private _IsCancelEnabled As Boolean = False
  Private _IsAddEnabled As Boolean = True
  Private _IsAddMode As Boolean = False
  Private _UIState As UIStateMode = UIStateMode.Normal
#End Region

#Region "Public Properties"
  Public Property IsSaveEnabled() As Boolean
    Get
      Return _IsSaveEnabled
    End Get
    Set(ByVal value As Boolean)
      If _IsSaveEnabled <> value Then
        _IsSaveEnabled = value
        RaisePropertyChanged("IsSaveEnabled")
      End If
    End Set
  End Property

  Public Property IsCancelEnabled() As Boolean
    Get
      Return _IsCancelEnabled
    End Get
    Set(ByVal value As Boolean)
      If _IsCancelEnabled <> value Then
        _IsCancelEnabled = value
        RaisePropertyChanged("IsCancelEnabled")
      End If
    End Set
  End Property

  Public Property IsAddEnabled() As Boolean
    Get
      Return _IsAddEnabled
    End Get
    Set(ByVal value As Boolean)
      If _IsAddEnabled <> value Then
        _IsAddEnabled = value
        RaisePropertyChanged("IsAddEnabled")
      End If
    End Set
  End Property

  Public Property IsAddMode() As Boolean
    Get
      Return _IsAddMode
    End Get
    Set(ByVal value As Boolean)
      If _IsAddMode <> value Then
        _IsAddMode = value
        RaisePropertyChanged("IsAddMode")
      End If
    End Set
  End Property

  Public Property UIState() As UIStateMode
    Get
      Return _UIState
    End Get
    Set(ByVal value As UIStateMode)
      If _UIState <> value Then
        _UIState = value
        SetViewStateMode()
        RaisePropertyChanged("UIState")
      End If
    End Set
  End Property
#End Region

#Region "SetViewStateMode Methods"
  Public Overridable Sub SetViewStateMode()
    SetViewStateMode(Me.UIState)
  End Sub

  Public Overridable Sub SetViewStateMode(ByVal mode As UIStateMode)
    UIState = mode
    Select Case mode
      Case UIStateMode.Normal
        IsAddMode = False
        IsCancelEnabled = False
        IsSaveEnabled = False
        IsAddEnabled = True
        Exit Select

      Case UIStateMode.Edit
        IsCancelEnabled = True
        IsSaveEnabled = True
        IsAddEnabled = False
        Exit Select

      Case UIStateMode.Add
        IsAddMode = True
        IsCancelEnabled = True
        IsSaveEnabled = True
        IsAddEnabled = False
        Exit Select

      Case UIStateMode.Cancel
        IsAddMode = False
        IsCancelEnabled = False
        IsSaveEnabled = False
        IsAddEnabled = True
        Exit Select
      Case Else

        Exit Select
    End Select
  End Sub
#End Region
End Class

File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples-2012/CSharp/WPFTemplateApp/WPFTemplateApp/Common/ViewModelBase.cs
namespace WPFTemplateApp
{
  public enum UIStateMode
  {
    Normal,
    Edit,
    Add,
    Cancel
  }

  public class ViewModelBase : ObjectBase
  {
    #region Private Variables
    private bool _IsSaveEnabled = false;
    private bool _IsCancelEnabled = false;
    private bool _IsAddEnabled = true;
    private bool _IsAddMode = false;
    private UIStateMode _UIState = UIStateMode.Normal;
    #endregion

    #region Public Properties
    public bool IsSaveEnabled
    {
      get { return _IsSaveEnabled; }
      set
      {
        if (_IsSaveEnabled != value)
        {
          _IsSaveEnabled = value;
          RaisePropertyChanged("IsSaveEnabled");
        }
      }
    }

    public bool IsCancelEnabled
    {
      get { return _IsCancelEnabled; }
      set
      {
        if (_IsCancelEnabled != value)
        {
          _IsCancelEnabled = value;
          RaisePropertyChanged("IsCancelEnabled");
        }
      }
    }

    public bool IsAddEnabled
    {
      get { return _IsAddEnabled; }
      set
      {
        if (_IsAddEnabled != value)
        {
          _IsAddEnabled = value;
          RaisePropertyChanged("IsAddEnabled");
        }
      }
    }

    public bool IsAddMode
    {
      get { return _IsAddMode; }
      set
      {
        if (_IsAddMode != value)
        {
          _IsAddMode = value;
          RaisePropertyChanged("IsAddMode");
        }
      }
    }
    
    public UIStateMode UIState
    {
      get { return _UIState; }
      set
      {
        if (_UIState != value)
        {
          _UIState = value;
          SetViewStateMode();
          RaisePropertyChanged("UIState");
        }
      }
    }
    #endregion

    #region SetViewStateMode Methods
    public virtual void SetViewStateMode()
    {
      SetViewStateMode(this.UIState);
    }

    public virtual void SetViewStateMode(UIStateMode mode)
    {
      UIState = mode;
      switch (mode)
      {
        case UIStateMode.Normal:
          IsAddMode = false;
          IsCancelEnabled = false;
          IsSaveEnabled = false;
          IsAddEnabled = true;
          break;

        case UIStateMode.Edit:
          IsCancelEnabled = true;
          IsSaveEnabled = true;
          IsAddEnabled = false;
          break;

        case UIStateMode.Add:
          IsAddMode = true;
          IsCancelEnabled = true;
          IsSaveEnabled = true;
          IsAddEnabled = false;
          break;

        case UIStateMode.Cancel:
          IsAddMode = false;
          IsCancelEnabled = false;
          IsSaveEnabled = false;
          IsAddEnabled = true;
          break;

        default:
          break;
      }
    }
    #endregion
  }
}
